examples/go/tagging/handler/handler.go

```diff
@@ -17,58 +17,56 @@ import (
 	"github.com/vmware/govmomi/vim25/types"
 )
 
-// tagInfo is part of the toml vcconfig file
-type tagInfo struct {
-	Action string
-	Urn    string
-}
-
-// VCenter is part of the toml vcconfig file
-type vCenterAuth struct {
-	Password string
-	Server   string
-	User     string
-}
-
 // Config represents the toml vcconfig file
 type vcConfig struct {
-	Tag     tagInfo
-	VCenter vCenterAuth
+	Tag struct {
+		Action string
+		Urn    string
+	}
+	VCenter struct {
+		Password string
+		Server   string
+		User     string
+	}
 }
 
-// The request from vmware-event-router.
-type dataObj struct {
-	Data nest1 `json:"data"`
+// The structure for a Cloud Event.
+type vmEvent struct {
+	Event event `json:"data"`
 }
-type nest1 struct {
-	Vm nest2 `json:"Vm"`
+type event struct {
+	Vm vmEventArgument `json:"Vm"`
 }
-type nest2 struct {
-	Vm vmInfo `json:"Vm"`
+type vmEventArgument struct {
+	Vm managedObjectReference `json:"Vm"`
 }
-type vmInfo struct {
-	Value string `json:"Value"`
+type managedObjectReference struct {
 	Type  string `json:"Type"`
+	Value string `json:"Value"`
 }
 
 // Handle a function invocation
 func Handle(req handler.Request) (handler.Response, error) {
 	ctx := context.Background()
 
 	// Retrieve vCenter and tag info from the config file.
-	vcc := loadTomlCfg()
+	cfg, err := loadTomlCfg()
+	if err != nil {
+		log.Error(err)
+		return
+	}
 
 	// Incoming request object.
-	var mor dataObj
-	err := json.Unmarshal(req.Body, &mor)
+	var event vmEvent
+	err := json.Unmarshal(req.Body, &event)
 	if err != nil {
 		return handler.Response{
 			Body:       nil,
 			StatusCode: http.StatusBadRequest,
 		}, err
 	}
 
-	if mor.Data.Vm.Vm.Value == "" || mor.Data.Vm.Vm.Type == "" {
+	if event.Event.Vm.Vm.Value == "" || event.Event.Vm.Vm.Type == "" {
```

Comment:
Consider writing a unit test for your fn and check whether this comparison is correct. It could be that instead of the empty default value `""` of a type you get `nil` which this check won't hit.

Reply:
I get an error `./handler.go:157:46: cannot convert nil to type types.ManagedObjectReference` when I try to compare `event.Data.Vm.Vm` to `nil`. Maybe we can cover this after you see the unit tests.

Reply:
Unit test for insufficient information for types.ManagedObjectReference was added. As far as I'm concerned, any event can trigger `gotag-fn` as long as it can send along VM value and VM type in the cloud event.